### PR TITLE
Coding best practices: `async` vs `await` improvements

### PR DESCRIPTION
**Pull request by henriqueholtz**
Async/await improvements according to [.NET AsyncGuidance](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md)

- Using `.GetAwaiter().GetResult()` instead of `.Result`;
- Getting rid some warnings;
- Turning the tests as async/await as possible. Getting rid of:
  - `.Wait()`
  - `.Result`
  - `.GetAwaiter().GetResult()`

Notes: 
- I know the PR became huge, but most changes are pretty similar thus easier to review.
- Some tests from `Ocelot.AcceptanceTests` were already failing (once in a while) at `develop` branch even before my changes, specially:
  - `KubernetesServiceDiscoveryTests.ShouldHighlyLoadOnUnstableKubeProvider_WithRoundRobinLoadBalancing`
  - `ConsulServiceDiscoveryTests.Should_send_request_to_service_after_it_becomes_available_in_consul`

### CODE REVIEW

**Comment by raman-m** [comment]
Hello, Henrique! Welcome to Ocelot!

Indeed, we have an issue with the improper use of threading operators, a legacy problem from the early days of Ocelot.
We are addressing it gradually. Your PR, however, has impressed me.

> Async/await improvements according to [.NET AsyncAwaitBestPractices](https://github.com/brminnick/AsyncAwaitBestPractices)

Are you aware that Microsoft Learn also provides documentation on best practices? I will find some links for you later.
For the moment, you need to resolve all merge conflicts.
**Have you rebased onto develop following the Oct 3 sync to main?**

**Comment by raman-m** [comment]
> - Some tests from `Ocelot.AcceptanceTests` were already failing (once in a while) at `develop` branch even before my changes, specially:
>   - `KubernetesServiceDiscoveryTests.ShouldHighlyLoadOnUnstableKubeProvider_WithRoundRobinLoadBalancing`
>   - `ConsulServiceDiscoveryTests.Should_send_request_to_service_after_it_becomes_available_in_consul`
> 

No worries, I will review them later.

**Comment by henriqueholtz** [comment]
> {...} Your PR, however, has impressed me.

Really ? Thanks for the feedback. I wasn't sure if it would be useful, specially because the PR's size.

> Are you aware that Microsoft Learn also provides documentation on best practices? I will find some links for you later. For the moment, you need to resolve all merge conflicts. 

I'm not sure, I'll be thankful to get these links.

> **Have you rebased onto develop following the Oct 3 sync to main?**

I only got based on `develop` branch. I'll resolve the conflicts and make sure the PR is up to date soon.

**Comment by henriqueholtz** [comment]
@raman-m as you can see I just solved the conflicts carefully. 

The PR's check is failing at `RunAcceptanceTests` step, however this error only occurs eventually. Test which is intermittently failing (see the image below):

- `ConsulServiceDiscoveryTests.ShouldUseConsulServiceDiscoveryAndLoadBalanceRequestWhenDynamicRoutingWithNoRoutes`

Note: Please let me know if you have any comments (good or not) about the changes, I'm really really interested on.

image

**Comment by raman-m** [comment]
Henrique, there's no need to be concerned about the failed tests. Indeed, we do have several unstable tests with a failure probability of 5-10% in CI-CD builds. They tend to be most unstable during US working hours when CircleCI experiences high traffic.
I've rerun the PR build, and it's passed now.
I am officially requesting a code review from the team...

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 1]
```diff
@@ -181,12 +184,23 @@ public async Task StartFakeOcelotWithWebSocketsWithConsul()
...
+    public async Task GivenThereIsAConfigurationAsync(FileConfiguration fileConfiguration)
+        => await GivenThereIsAConfigurationAsync(fileConfiguration, _ocelotConfigFileName);
+
+    public async Task GivenThereIsAConfigurationAsync(FileConfiguration from, string toFile)
+    {
+        toFile ??= _ocelotConfigFileName;
+        var jsonConfiguration = JsonConvert.SerializeObject(from, Formatting.Indented);
+        await File.WriteAllTextAsync(toFile, jsonConfiguration);
+        Files.Add(toFile); // register for disposing
+    }
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:197]: Simply return the `Task` object as it can be refactored in this manner 👇 
```suggestion
    public Task GivenThereIsAConfigurationAsync(FileConfiguration from, string toFile)
    {
        toFile ??= _ocelotConfigFileName;
        Files.Add(toFile); // register for disposing
        var jsonConfiguration = JsonConvert.SerializeObject(from, Formatting.Indented);
        return File.WriteAllTextAsync(toFile, jsonConfiguration);
    }
```

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 2]
```diff
@@ -207,26 +221,26 @@ protected virtual void DeleteFiles()
...
+        foreach (var folder in Folders)
+        {
+            try
+            {
+                var f = new DirectoryInfo(folder);
+                if (f.Exists && f.FullName != AppContext.BaseDirectory)
+                {
+                    f.Delete(true);
+                }
+            }
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:238]: Ugly fake changes! 🙄

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 3]
```diff
@@ -275,6 +289,62 @@ protected void StartOcelot(Action<WebHostBuilderContext, IConfigurationBuilder>
         _ocelotClient = _ocelotServer.CreateClient();
     }
 
+    /// <summary>
+    /// This is annoying cos it should be in the constructor but we need to set up the file before calling startup so its a step.
+    /// </summary>
+    /// <typeparam name="T">The <see cref="ILoadBalancer"/> type.</typeparam>
+    /// <param name="loadBalancerFactoryFunc">The delegate object to load balancer factory.</param>
+    public void GivenOcelotIsRunningWithCustomLoadBalancer<T>(Func<IServiceProvider, DownstreamRoute, IServiceDiscoveryProvider, T> loadBalancerFactoryFunc)
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:297]: Why was this method reintroduced? It should have no references as it is no longer utilized.
### Remove it❗

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 4]
```diff
@@ -275,6 +289,62 @@ protected void StartOcelot(Action<WebHostBuilderContext, IConfigurationBuilder>
...
+                s.AddOcelot()
+                    .AddCustomLoadBalancer(loadBalancerFactoryFunc);
+            })
+            .Configure(async app => { await app.UseOcelot(); });
+
+        _ocelotServer = new TestServer(_webHostBuilder);
+        _ocelotClient = _ocelotServer.CreateClient();
+    }
+
+    public void GivenOcelotIsRunningWithConsul(params string[] urlsToListenOn)
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:321]: It appears to have no references, correct? We have ceased using this method in favor of the `GivenOcelotIsRunningWithServices` method.
### Please remove it❕

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 5]
```diff
@@ -345,7 +415,11 @@ public void GivenOcelotIsRunningUsingConsulToStoreConfigAndJsonSerializedCache()
                     .AddConsul()
                     .AddConfigStoredInConsul();
             })
-            .Configure(app => { app.UseOcelot().Wait(); });
+            .Configure(app =>
+            {
+                // Turning as async/await some tests got broken
+                app.UseOcelot().GetAwaiter().GetResult();
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:421]: Why not await ?

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 6]
```diff
@@ -381,7 +459,7 @@ public void WhenIGetUrlOnTheApiGatewayWaitingForTheResponseToBeOk(string url)
         {
             try
             {
-                _response = _ocelotClient.GetAsync(url).Result;
+                _response = _ocelotClient.GetAsync(url).GetAwaiter().GetResult();
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:462]: Did you try to convert the method to `async`? And after that ->
```suggestion
                _response = await _ocelotClient.GetAsync(url);
```

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 7]
```diff
@@ -566,38 +644,38 @@ public void GivenOcelotIsRunningWithGlobalHandlersRegisteredInDi<TOne>(FakeDepen
...
-
+    public async Task WhenIGetUrlOnTheApiGatewayWithCookieAsync(string url, string cookie, string value)
+        => _response = await WhenIGetUrlOnTheApiGatewayAsync(url, cookie, value);
+    public async Task WhenIGetUrlOnTheApiGatewayWithCookieAsync(string url, CookieHeaderValue cookie)
+        => _response = await WhenIGetUrlOnTheApiGatewayAsync(url, cookie);
+
+    public async Task<HttpResponseMessage> WhenIGetUrlOnTheApiGatewayAsync(string url, string cookie, string value)
+    {
+        var header = new CookieHeaderValue(cookie, value);
+        return await WhenIGetUrlOnTheApiGatewayAsync(url, header);
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:667]: Why do you await here? 🙄

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 8]
```diff
@@ -784,84 +862,85 @@ public void GivenOcelotIsRunningWithEureka()
...
+        content.Add(dataContent, name);
+        content.Headers.ContentDisposition = new ContentDispositionHeaderValue("form-data");
 
         var request = new HttpRequestMessage(HttpMethod.Get, url)
         {
             Content = content,
-        };
-        _response = _ocelotClient.SendAsync(request).Result;
+        };
+        _response = _ocelotClient.SendAsync(request).GetAwaiter().GetResult();
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:891]:
```suggestion
        _response = await _ocelotClient.SendAsync(request);
```

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 9]
```diff
@@ -891,25 +970,33 @@ public void GivenThePostHasGzipContent(object input)
...
-        => _response.StatusCode.ShouldBe(expected);
-
     public void ThenTheStatusCodeShouldBe(int expectedHttpStatusCode)
     {
         var responseStatusCode = (int)_response.StatusCode;
         responseStatusCode.ShouldBe(expectedHttpStatusCode);
     }
 
+    public void ThenTheStatusCodeShouldBe(HttpStatusCode expected)
+        => _response.StatusCode.ShouldBe(expected);
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:997]: Can't get this movement!

File: test/Ocelot.Benchmarks/AllTheThingsBenchmarks.cs [hunk 10]
```diff
@@ -99,9 +99,9 @@ private void GivenOcelotIsRunning(string url)
                     logging.AddConfiguration(hostingContext.Configuration.GetSection("Logging"));
                 })
                 .UseIISIntegration()
-                .Configure(app =>
+                .Configure(async app =>
                 {
-                    app.UseOcelot().Wait();
+                    await app.UseOcelot();
```
Comment (raman-m) [inline test/Ocelot.Benchmarks/AllTheThingsBenchmarks.cs:104]: `WithUseOcelot` ?
I mean &rarr; `.Configure(WithUseOcelot)`

File: test/Ocelot.IntegrationTests/CacheManagerTests.cs [hunk 11]
```diff
@@ -1,16 +1,16 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Newtonsoft.Json;
+using Ocelot.Administration;
+using Ocelot.Cache.CacheManager;
+using Ocelot.Configuration.File;
+using Ocelot.DependencyInjection;
+using Ocelot.Middleware;
+using System.Net;
+using System.Net.Http.Headers;
```
Comment (raman-m) [inline test/Ocelot.IntegrationTests/CacheManagerTests.cs:13]: 😠

**Review by raman-m — CHANGES_REQUESTED** [review]
The primary issue with this PR appears to be the artificial changes resulting from an EOL discrepancy. For more details, please refer to the initial code review comment.

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 12]
```diff
@@ -181,12 +184,23 @@ public async Task StartFakeOcelotWithWebSocketsWithConsul()
 
     public void GivenThereIsAConfiguration(FileConfiguration fileConfiguration)
         => GivenThereIsAConfiguration(fileConfiguration, _ocelotConfigFileName);
-
+
+    public async Task GivenThereIsAConfigurationAsync(FileConfiguration fileConfiguration)
+        => await GivenThereIsAConfigurationAsync(fileConfiguration, _ocelotConfigFileName);
```
[the remark this answers is not in the document]
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs:189]: I disagree. Are you sure about it? 

Take a look at the section [Prefer async/await over directly returning Task](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task)

Note: I updated the reference link at the MR's summary because it was the wrong one.

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 13]
```diff
@@ -56,30 +59,30 @@ public Steps()
...
-        UpstreamHttpMethod = new() { HttpMethods.Get },
+    };
+
+    protected static FileRoute GivenDefaultRoute(int port) => new()
+    {
+        DownstreamPathTemplate = "/",
+        DownstreamHostAndPorts = new() { Localhost(port) },
+        DownstreamScheme = Uri.UriSchemeHttp,
+        UpstreamPathTemplate = "/",
+        UpstreamHttpMethod = new() { HttpMethods.Get },
```
[the remark this answers is not in the document]
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs:85]: I'm using Visual Studio Community 2022, and I haven't overridden the `.gitattributes` neither `.editorconfig`. Should I install a specific extension?

I'm really interested on be within the rules of the project, but I'm not sure what's lacking (BTW, thanks for letting me know you're dissatisfied with that, I would be the same way).

File: test/Ocelot.AcceptanceTests/Steps.cs
[the remark this answers is not in the document]
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs:1]: Good point. I want to finish this MR as it's and hopefully raise these changes at a new MR. Ok? 

The main reason is that this MR's already huge.

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 2 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs:238]: I hope these changes are related to [this comment](https://github.com/ThreeMammals/Ocelot/pull/2156/files/bca0ba0edc6ca744c65bbebe8ae1a675a104dab4#r1792202279) because I don't remember doing that.

File: test/Ocelot.Benchmarks/AllTheThingsBenchmarks.cs (hunk 10 above)
Comment (henriqueholtz) [reply test/Ocelot.Benchmarks/AllTheThingsBenchmarks.cs:104]: The method to do that only exists at `Ocelot.AcceptanceTests.Steps`. Should I copy the method `WithUseOcelot` to `Ocelot.Benchmarks.AllTheThingsBenchmarks` (or even add reference to the `Ocelot.AcceptanceTests`) ? 

I can't see any benefit doing that. But I'm open to do if you prefer, just let me know.

<details><summary>Example as async</summary>

image

</details> 


<details><summary>Example as sync</summary>

Note: Using `.Wait()` or `.GetAwaiter().GetResult` 

image


</details>

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 5 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs:421]: The test `should_return_response_200_with_simple_url_when_using_jsonserialized_cache` (at `Ocelot.AcceptanceTests.Configuration.ConfigurationInConsulTests` starts to fail. I'm not sure why.

image

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 1 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs]: Done!

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 3 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs]: Done!

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 4 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs]: Done!

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 6 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs]: I tried but I couldn't. I retried and now I could came up with the async version for `Until` which is, of course, called `UntilAsync`. This way we got rid some `GetAwaiter().GetResult()`.

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 7 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs]: I realize I misunderstand this part at the async guide at [Prefer async/await over directly returning Task](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task). I agree we should just return the `Task`. 

Reading the piece below... Is it looks confuse for you too?

image

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 9 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs]: Undone!

File: test/Ocelot.IntegrationTests/CacheManagerTests.cs (hunk 11 above)
Comment (henriqueholtz) [reply test/Ocelot.IntegrationTests/CacheManagerTests.cs:13]: Hopefully we'll solve it [here](https://github.com/ThreeMammals/Ocelot/pull/2156#discussion_r1792202279)

**Review by RaynaldM — APPROVED** [review]

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs:85]: It is hard to read PR changes having fake changes!
It's preferable to resolve merge conflicts while honoring the changes in the develop branch. It appears that changes are being collected from the feature branch, even when there are no substantial changes. However, conflicts should be resolved by applying your changes onto the develop branch using a merging tool.

If changes from the feature branch are prioritized (despite being insignificant), the merge tool will record them and apply CRLF end-of-line characters based on the rules specified in `.editorconfig`. This is where the issue arises.

File: test/Ocelot.AcceptanceTests/Steps.cs
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs:1]: Do whatever you want.
The PR is huge because of fake changes.
I'd suggest to focus on improvements in Ocelot src folder ignoring testing projects.

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 2 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs:238]: The link is broken! It goes nowhere!

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 3 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs]: If it is done, feel free to press "Resolve" button. If you want to highlight resolving commit then write a hash of the commit.

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 5 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs:365]: Ah, it should wait synchronously to setup app pipeline... Yes, this is current bottleneck in the design.
Not an issue.

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 6 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs]: Seems this is special case of the test. So, the test must wait synchronously...
Not an issue!

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 7 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs]: I can't dispute Mr. Fowler's point, and his advice is indeed beneficial for the library code, not the testing project. Performance is my priority during testing: tests need to be quick. I'm not concerned about leaking internal variables into the broader context in testing projects or a disrupted call stack.

Please fix this!

File: test/Ocelot.Benchmarks/AllTheThingsBenchmarks.cs (hunk 10 above)
Comment (raman-m) [reply test/Ocelot.Benchmarks/AllTheThingsBenchmarks.cs:104]: Oh, sorry! This is benchmarks project. Not an issue!
Don't copy!
We have common testing project: [Ocelot.Testing](https://github.com/ThreeMammals/Ocelot/tree/develop/test/Ocelot.Testing)

File: test/Ocelot.IntegrationTests/CacheManagerTests.cs (hunk 11 above)
Comment (raman-m) [reply test/Ocelot.IntegrationTests/CacheManagerTests.cs:13]: Probably! Seems you need to rebase once again 😉

**Comment by raman-m** [comment]
@RaynaldM I believe this PR is not [ready](https://github.com/ThreeMammals/Ocelot/pull/2156#pullrequestreview-2359221689) to be merged!

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 7 above)
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/Steps.cs]: Already did!

**Comment by raman-m** [comment]
Hi, what have you [decided](https://github.com/ThreeMammals/Ocelot/pull/2156#discussion_r1793350677) about this PR? Will you keep this PR open or create a new one?

**Comment by henriqueholtz** [comment]
@raman-m I carefully improved the PR removing so many fake changes. There were 92 changed files before and now we have 60. Also into there 60 they have lesser changes in each (because I reverted the renaming adding the suffix `Async`).

Reverting fake changes was successfully done for most files using commands like below (`origin` in my case is pointing to original repository):

```
git checkout origin/develop -- test/Ocelot.AcceptanceTests/ButterflyTracingTests.cs
```

However I couldn't be able to get rid some fake changes at some files which has those fake changes but also some real changes like `test/Ocelot.UnitTests/CacheManager/OutputCacheMiddlewareRealCacheTests.cs`:

I also tried to revert and manually re-add the real changes in VS but VS Code and didn't work which looks to be a problem with some repository's config (probably at `.editorconfig`). I tried add some like `eol=lf` but didn't work as well.

Example:

image

Note: I could got rid the above fake changes doing: reverting all the changes using the command above and bringing changes manually using `notepad++` . I know it's a bad idea but I did want to be sure if there is a problem only using the IDEs. You can see the attempts at PR's commit history.

**I turned the PR as ready to review again with a few files at this scenario.** If you want to try something feel free to ask or even to do directly (and let me know please).

Files which has some few fake changes yet:

- `test/Ocelot.UnitTests/Configuration/FileConfigurationSetterTests.cs`
- `test/Ocelot.UnitTests/Requester/HttpRequesterMiddlewareTests.cs`
- `test/Ocelot.UnitTests/Security/SecurityMiddlewareTests.cs`
- `test/Ocelot.UnitTests/Configuration/Validation/FileConfigurationFluentValidatorTests.cs`
- `test/Ocelot.IntegrationTests/HeaderTests.cs`
- `test/Ocelot.IntegrationTests/CacheManagerTests.cs`
- `test/Ocelot.AcceptanceTests/Steps.cs`
- `test/Ocelot.AcceptanceTests/AggregateTests.cs`

**Comment by raman-m** [comment]
Henrique, you made a mistake before the creation of this PR →
> feature branch: [henriqueholtz:develop](https://github.com/henriqueholtz/Ocelot/tree/develop)

The best practice is to always create a separate feature branch from develop, since develop is the default branch of the fork that needs to be synchronized with the upstream repository.

Additional commits in the default 'develop' branch can restrict actions such as rebasing and creating new branches. Rebasing, in particular, becomes impossible, leaving only manual work as a viable option. 🤔

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 14]
```diff
@@ -56,7 +56,7 @@ private static async Task CopyAsync(Stream input, Stream output, long announcedC
                 if (zeroByteReadTask.IsCompletedSuccessfully)
                 {
                     // Consume the ValueTask's result in case it is backed by an IValueTaskSource
-                    _ = zeroByteReadTask.Result;
+                    _ = zeroByteReadTask.GetAwaiter().GetResult();
```
Comment (raman-m) [inline src/Ocelot/Request/Mapper/StreamHttpContent.cs:59]: It appears that reading `Result` once is acceptable when the value task is complete, as indicated by the [ValueTask{TResult}.Result Property | Remarks](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1.result?view=net-8.0#remarks). Therefore, this change may be unnecessary due to the assurance provided by `.IsCompletedSuccessfully`.
@ggnaegi, would you agree?

File: test/Ocelot.AcceptanceTests/AggregateTests.cs
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/AggregateTests.cs:1]: Oh, my gush! 🤯  👇 700 lines of fake changes!
Can't you make manual changes?

**Review by raman-m — CHANGES_REQUESTED** [review]
I'm uncertain who will correct the fictitious changes. Such tasks will consume our precious time...

File: test/Ocelot.AcceptanceTests/AggregateTests.cs
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/AggregateTests.cs:1]: Of course!

File: src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs [hunk 15]
```diff
@@ -18,7 +18,7 @@ private int GetDelay()
         {
             var delay = 1000;
 
-            var fileConfig = Task.Run(async () => await _fileConfigurationRepository.Get()).Result;
+            var fileConfig = Task.Run(async () => await _fileConfigurationRepository.Get()).GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (henriqueholtz) [reply src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs]: I saw that and thought the same. However as it's my first PR at this repo (and as you already noticed, one of my few first), then some changes I just skipped. In this case I just pushed the change removing the useless `Task`.

[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs]: In this instance, minor code refactoring is advisable. Please proceed with the modification.

[the remark this answers is not in the document]
Comment (henriqueholtz) [reply src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs]: Done yesterday!

File: test/Ocelot.AcceptanceTests/AggregateTests.cs
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/AggregateTests.cs:1]: @raman-m I was trying to figure out why those "fake changes" even though we have the `.editorconfig` in place. 

Note: the right version is `CRLF` (as it's in `.editorconfig` file). 

Those "fake changes" are because some lines at these files are still as `LF` (incorrect!). It means that these changes are actually the correct way to do (I know it let the PR longer/harder to review).

Are you aware of that? Should I undo it even though ? It's important to point out that any person who touch these files using an IDE like VS or VS Code will hit the same problem. 

Basically if I undo these "fake changes", when someone else touch these files they'll face the same issue. Considering the long way to the repo please consider that we should keep the changes as the `.editorconfig` say. Take a look at the example below. 

Also you can just try to do that on your machine and see/confirm it.

Example `test/Ocelot.UnitTests/Configuration/FileConfigurationSetterTests.cs`:

image

Comment (raman-m) [reply test/Ocelot.AcceptanceTests/AggregateTests.cs:1]: > I was trying to figure out why those "fake changes" even though we have the .editorconfig in place.

I've explained the root cause of fake changes [here](https://github.com/ThreeMammals/Ocelot/pull/2156#issuecomment-2405053138): this is auto-refactoring commands in IDE (Visual Studio, Visual Code, Raider and etc.)

> Those "fake changes" are because some lines at these files are still as LF (incorrect!). It means that these changes are actually the correct way to do (I know it let the PR longer/harder to review).

This issue has persisted since the project's inception in 2016!
Indeed, some lines end with the `LF` character from the Linux OS. Several of our contributors work on Linux and use IDEs like Visual Code, where the default newline character is `LF`. Consequently, we have numerous files with inconsistent EOL characters. Do you code on Linux or Windows?
**I suggest using Windows to avoid the end-of-line (EOL) issue. ❗**

> Are you aware of that? Should I undo it even though ? It's important to point out that any person who touch these files using an IDE like VS or VS Code will hit the same problem.

Yes, I am! What are you trying to teach me, man? I've been aware of this issue for over a year.
You'll encounter the problem if you use Linux OS.
**Boot into Windows, use Visual Studio Community (it's free), avoid using auto-refactoring commands, and everything will work out fine.**

>Basically if I undo these "fake changes", **when someone else touch these files they'll face the same issue.** Considering the long way to the repo please consider that we should keep the changes as the .editorconfig say. Take a look at the example below.

It's irrelevant! The likelihood of encountering this issue remains unchanged. We cannot foresee which operating system the subsequent developer might use. The concern is not regarding future developers, but rather about your current development environment.
If your PR contains numerous artificial changes aka "fake changes", it increases the chances that subsequent developers will encounter this issue. Therefore, I request the removal of all such changes due to end-of-line characters.

**Comment by raman-m** [comment]
@henriqueholtz I'm initiating a rebase of develop onto upstream develop...
Please refrain from pushing/pulling during this process❗ The rebase will address any inconsistencies in the branch, ensuring that all new commits are placed on top of the upstream develop commits.

File: src/Ocelot.Provider.Consul/Consul.cs [hunk 16]
```diff
@@ -30,9 +30,8 @@ public virtual async Task<List<Service>> GetAsync()
         var nodesTask = _consul.Catalog.Nodes();
 
         await Task.WhenAll(entriesTask, nodesTask);
-
-        var entries = entriesTask.Result.Response ?? Array.Empty<ServiceEntry>();
-        var nodes = nodesTask.Result.Response ?? Array.Empty<Node>();
+        var entries = (await entriesTask).Response ?? Array.Empty<ServiceEntry>();
+        var nodes = (await nodesTask).Response ?? Array.Empty<Node>();
```
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot.Provider.Consul/Consul.cs:34]: Will we await 3 times?
The goal of `Task.WhenAll` is parallelization!

File: src/Ocelot.Provider.Consul/PollConsul.cs [hunk 17]
```diff
@@ -51,7 +51,7 @@ public Task<List<Service>> GetAsync()
             try
             {
                 _logger.LogInformation(() => $"Retrieving new client information for service: {ServiceName}...");
-                _services = _consulServiceDiscoveryProvider.GetAsync().Result;
+                _services = _consulServiceDiscoveryProvider.GetAsync().GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot.Provider.Consul/PollConsul.cs:54]: Not an issue.

File: src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs [hunk 18]
```diff
@@ -18,7 +18,7 @@ private int GetDelay()
         {
             var delay = 1000;
 
-            var fileConfig = Task.Run(async () => await _fileConfigurationRepository.Get()).Result;
+            var fileConfig = _fileConfigurationRepository.Get().GetAwaiter().GetResult();
```
Comment (raman-m) [inline src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs:21]: We have this sync version in the interface:
https://github.com/ThreeMammals/Ocelot/blob/0379c7778f05b77aeea8c5e9f4512a1ea2e30ce8/src/Ocelot/Configuration/Repository/IFileConfigurationPollerOptions.cs#L5
Why not to add async version to the interface?
After that we can make private and public methods as true async &rarr; `Task<int> DelayAsync { get; }`
```suggestion
            var fileConfig = await _fileConfigurationRepository.Get();
```
Optional goal is using `Task.Delay`, if applicable.

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs (hunk 14 above)
Comment (raman-m) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:59]: This change is not required. See details above.

**Review by raman-m — CHANGES_REQUESTED** [review]
Going to fix "fake changes"...

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs (hunk 14 above)
Comment (raman-m) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs]: Fixed in https://github.com/ThreeMammals/Ocelot/pull/2156/commits/8b107922ce3aa4e6941ea21c0ff7bc62912e6507

File: src/Ocelot.Provider.Consul/Consul.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot.Provider.Consul/Consul.cs:34]: I've double checked the MS docs and both variants are good
- we can read already initialized `Result` property because the task was completed by `Task.WhenAll`
- we can await twice with `await task`, no new execution because of task completion (see [remarks](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task-1.result?view=net-8.0#remarks)), and read the result from implicit local variable in expression `(await task).prop`

The 2nd statement has better readability, that's why **there is no issue.**

File: test/Ocelot.AcceptanceTests/AggregateTests.cs
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/AggregateTests.cs:1]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2156/commits/a851098b68617b1d14821adc6d7dc913024c5a90 ✔️

File: src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs (hunk 18 above)
Comment (raman-m) [reply src/Ocelot/Configuration/Repository/ConsulFileConfigurationPollerOption.cs:21]: The expression has been revised to execute synchronously within the synchronous `int GetDelay()` method.
The proposed suggestion for refactoring is beyond the scope of this pull request. We will refactor the interface in future.

**Review by raman-m — APPROVED** [review]
### Ready for delivery ✅

**Comment by raman-m** [comment]
### P.S.
Please re-fork the Ocelot repository into your account due to the previously [mentioned issue](https://github.com/ThreeMammals/Ocelot/pull/2156#issuecomment-2408992027) of having extra commits in the develop branch, which should only contain commits from the upstream develop.

You will need to re-fork the repository.

File: test/Ocelot.AcceptanceTests/AggregateTests.cs
Comment (henriqueholtz) [reply test/Ocelot.AcceptanceTests/AggregateTests.cs:1]: Got the same issue using Windows (VS and VS Code), but also using linux. That's because I asked you gently, while I was trying to help. You was totally rough with no reason to act like that.
